### PR TITLE
[#1] [Chore] Set up Initial project

### LINKED ISSUE #1: [Chore] Set up Initial project
## Why

Set up the Initial project to begin work, includes:
- Create a new project based on [gin-templates](https://github.com/nimblehq/gin-templates)

## Who Benefits?

Developer

### PR DESCRIPTION
**Pull request by markgravity**
- Close #1

## What happened 👀

- Create a new project based on [gin-templates](https://github.com/nimblehq/gin-templates)

## Insight 📝

- Using this cmd:
```
go install github.com/nimblehq/gin-templates@latest
gin-templates create
```
- Known issue:
`make dev`

```
[+] Running 2/2
 ✔ Network mark-golang-ic_default  Created                                                                                                                                                       0.0s 
 ✔ Container mark-ic_db            Started                                                                                                                                                       0.3s 
make db/migrate
make wait-for-postgres
Postgres is unavailable - sleeping
Postgres is unavailable - sleeping
```
CleanShot 2023-08-14 at 15 01 41

- The CI will be set up at #2, so please ignore the error now

## Proof Of Work 📹

n/a

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+migration/status:
+	goose -dir database/migrations -table "migration_versions" postgres "$(DATABASE_URL)" status
+
+dev:
+	make env-setup
+	make db/migrate
+	forego start -f Procfile.dev
+
+install-dependencies:
+	go install github.com/cosmtrek/air@v1.41.1
```
Comment (malparty) [inline Makefile]:
```suggestion
	go install github.com/cosmtrek/air@v1.44.0
```

I had an error message telling me it cannot find the v1.41.1 🤔

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Resolve https://github.com/nimblehq/git-template/issues/??
```
Comment (malparty) [inline .github/PULL_REQUEST_TEMPLATE.md]: I recommend changing these directly, so you wont' have to do it later 😇 (same for all the templates :))

You can set it like the way you used in this PR 🚀 

```
- close #
```

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -0,0 +1 @@
+* @suho @malparty
```
Comment (malparty) [inline .github/CODEOWNERS]: Let's add `markgravity` as well, his contributions will help the project move forward 🦄  

...and maybe also an empty line to end the file 😇

File: README.md [hunk 4]
```diff
@@ -1 +1,132 @@
-# golang-ic
+[![Build Status](CI_BADGE_URL goes here)](REPO_URL goes here)
+
+## Introduction
+
+> *App introduction goes here ...*
```
Comment (malparty) [inline README.md]: Let's directly have the app intro filled, otherwise we might forget about it :)

File: database/database.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+var database *gorm.DB
+
+func InitDatabase(databaseURL string) {
+	db, err := gorm.Open(postgres.Open(databaseURL), &gorm.Config{})
+	if err != nil {
+		log.Fatalf("Failed to connect to %v database: %v", gin.Mode(), err)
+	} else {
+		viper.Set("database", db)
+		log.Println(strings.Title(gin.Mode()) + " database connected successfully.")
+	}
```
Comment (malparty) [inline database/database.go:25]: Interesting suggestion from GitHub, we might want to open a quick and easy fix in the Gin-template to replace this `Title` call 😇

File: test/test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+func SetupTestEnvironment() {
+	gin.SetMode(gin.TestMode)
+
+	setRootDir()
+
+	bootstrap.LoadConfig()
+
+	bootstrap.InitDatabase()
+}
```
Comment (malparty) [inline test/test.go:25]: Same `not enough arguments in call to bootstrap.InitDatabase` error on my machine, I recommend fixing that before merging this PR (it's probably due to a package update that changed the way the DB is initialized)

**Review by malparty — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+on:
+  workflow_run:
+    workflows:
+      - Test
+    branches:
+      - main
+      - master
+      - develop
+      - development
```
Comment (suho) [inline .github/workflows/deploy.yml]: Based on our convention, now you just need 2 branches
```suggestion
      - main
      - develop
```

File: .github/workflows/deploy.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    types:
+      - completed
+  workflow_dispatch:
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Cancel previous runs
+        uses: styfle/cancel-workflow-action@0.9.1
```
Comment (suho) [inline .github/workflows/deploy.yml]: For cancel previous runs, we could use the native one from GitHub (https://docs.github.com/en/actions/using-jobs/using-concurrency), don't need to use a custom. You could create a new issue for this change

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
+name: Test
+
+on: push
```
Comment (suho) [inline .github/workflows/test.yml]: We should apply test for `pull_request` event, `push` is unnecessary

File: README.md [hunk 10]
```diff
@@ -1 +1,132 @@
...
+
+## Project Setup
+
+### Prerequisites
+
+- [Go - 1.20](https://golang.org/doc/go1.20) or newer
+
+
+
```
Comment (suho) [inline README.md]: This might be the problem from the templates, but it's better to remove these extra lines

**Comment by markgravity** [comment]
@malparty solved your comments 🙏 
[5963d8c](https://github.com/markgravity/golang-ic/pull/18/commits/5963d8c6a2522141cea9e2311c38220800cc1487)

File: test/test.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
+package test
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/nimblehq/mark-ic/bootstrap"
```
Comment (suho) [inline test/test.go]: You might need to use the right repo URL for importing

```suggestion
	"github.com/markgravity/golang-ic/bootstrap"
```

Comment (markgravity) [reply test/test.go]: @suho it's weird the current import is correct 🤔 
CleanShot 2023-08-15 at 10 22 11

**Comment by markgravity** [comment]
@suho resolved your comments here
[716f5bb](https://github.com/markgravity/golang-ic/pull/18/commits/716f5bb56e1aa23c34c5d95970de5a0ca95dd28e)

File: test/test.go (hunk 11 above)
Comment (suho) [reply test/test.go]: @markgravity This is defined in the `go.mod` file

Comment (markgravity) [reply test/test.go]: @suho Gotcha, updated at [437555d](https://github.com/markgravity/golang-ic/pull/18/commits/437555d9ac6a64630506bc5bc13272c8874e7cff)

File: .github/CODEOWNERS [hunk 12]
```diff
@@ -0,0 +1 @@
+* @suho @malparty @markgravity
```
Comment (suho) [inline .github/CODEOWNERS]: Please add a new line at the end of this file

**Review by suho — APPROVED** [review]

**Review by malparty — APPROVED** [review]
Approved ahead with the last comment about missing ending emply line.

**Pull request merged**